modules/packages/Sort.chpl

```diff
@@ -1233,52 +1233,62 @@ module QuickSort {
     import Sort.InsertionSort;
 
     // grab obvious indices
-    const lo = start,
-          hi = end,
-          mid = lo + (hi-lo+1)/2;
-    var piv = mid;
-
-    if hi - lo < 0 { // minlen {
-      // base case -- use insertion sort
-      InsertionSort.insertionSortMoveElts(Data, comparator=comparator, lo, hi);
-      return;
-    } else if hi <= lo {
-      // nothing to sort
-      return;
-    }
+    var lo = start,
+          hi = end;
 
-    // find pivot using median-of-3 method for small arrays
-    // and a "ninther" for bigger arrays. Places the pivot in
-    // Data[lo].
-    if hi - lo < 100 {
-      piv = order3(Data, lo, mid, hi, comparator);
-    } else {
-      // assumes array size > 9 at the very least
 
-      // median of each group of 3
-      const medLo  = order3(Data, lo,    lo+1, lo+2,  comparator);
-      const medMid = order3(Data, mid-1, mid,  mid+1, comparator);
-      const medHi  = order3(Data, hi-2,  hi-1, hi,    comparator);
-      // median of the medians
-      piv = order3(Data, medLo, medMid, medHi, comparator);
-    }
+    while(lo <= hi){
+        var mid = lo + (hi-lo+1)/2;
+        var piv = mid;
 
-    var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+        if hi - lo < 0 { // minlen {
+          // base case -- use insertion sort
+          InsertionSort.insertionSortMoveElts(Data, comparator=comparator, lo, hi);
+          return;
+        } else if hi <= lo {
+          // nothing to sort
+          return;
+        }
 
-    if hi-lo < 300 {
-      // stay sequential
-      quickSortImpl(Data, minlen, comparator, lo, eqStart-1);
-      quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
-    } else {
-      // do the subproblems in parallel
-      forall i in 1..2 {
-        if i == 1 then
-          quickSortImpl(Data, minlen, comparator, lo, eqStart-1);
-        else
-          quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+        // find pivot using median-of-3 method for small arrays
+        // and a "ninther" for bigger arrays. Places the pivot in
+        // Data[lo].
+        if hi - lo < 100 {
+          piv = order3(Data, lo, mid, hi, comparator);
+        } else {
+          // assumes array size > 9 at the very least
+
+          // median of each group of 3
+          const medLo  = order3(Data, lo,    lo+1, lo+2,  comparator);
+          const medMid = order3(Data, mid-1, mid,  mid+1, comparator);
+          const medHi  = order3(Data, hi-2,  hi-1, hi,    comparator);
+          // median of the medians
+          piv = order3(Data, medLo, medMid, medHi, comparator);
+        }
+
+        var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+
+        if hi-lo < 300 {
+          // stay sequential
+          if(eqStart - lo > hi - eqEnd){
+            quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+            hi = eqStart - 1;
+          } else {
```

Comment:
Now that this code has an `if`/`else` it is only doing half of the required quick sorting. It is my understanding that implementations of quick sort that do not use a recursive call need to have a manual "stack" of regions of the array that need to be sorted.

I don't think that what you have here is not doing tail call optimization. Rather it is changing the algorithm. Is there some prior work describing why this algorithm change would be a good idea? It does not look right to me.

Certainly it is possible to directly do one of the two subproblems and put the other one on a stack for later work, but that's not what this does.

I would expect that if we change it to use a stack of subproblems, it could do things like sort that stack of subproblems by size in order to have better parallel load balance.

Reply:
>Now that this code has an if/else it is only doing half of the required quick sorting.

I ran a few tests, it actually completes both subproblems but the larger half is solved in the current task/function call only whereas for smaller we make a recursive call the role of `while (lo <= hi)` and `hi = eqStart - 1` or `lo = eqEnd + 1` ensures that this work is performed for the left out the problem as well. 

I wasn't able to find any resource for tail call elimination for three-way quick sort so I carried over the idea from two way quicksort from the following [resource](https://www.geeksforgeeks.org/quicksort-tail-call-optimization-reducing-worst-case-space-log-n/)

>It is my understanding that implementations of quicksort that do not use a recursive call need to have a manual "stack" of regions of the array that need to be sorted.

The stack ones that you mentioned we can use but will be a completely iterative solution which will be more complex to implement as per a paper I read about load balancing in parallel but performance gain will be good but currently, I was just trying to solve just the tail call optimization problem.

Reply:
Ah, I see in your linked resource, a comment like this:

>         // If left part is smaller, then recur for left
>         // part and handle right part iteratively

This is the part that I was missing. I think what you have is probably fine but we need to add a comment about how both the left and right parts are handled; one iteratively and one in a recursive call. I don't see any problem extending this idea to the 3-way quicksort (we just leave the equal elements alone).

Reply:
Okay I will add few more documentation comments to make the idea more clear